File: packages/shared_types/transactions.py
from datetime import datetime
from enum import Enum
from typing import Optional, Any

from pydantic import BaseModel, Field, field_validator


class TransactionDirection(str, Enum):
    DEBIT = "DEBIT"
    CREDIT = "CREDIT"


class TransactionStatus(str, Enum):
    PENDING = "PENDING"
    POSTED = "POSTED"
    REVERSED = "REVERSED"
    CANCELLED = "CANCELLED"


class TransactionResponse(BaseModel):
    """
    Public-safe version of a Transaction.
    Returned by API endpoints (e.g. case detail, alerts).
    """

    id: int
    tx_id: str
    account_id: str
    customer_id: Optional[str] = None

    invoice_id: Optional[int] = None

    tx_time: datetime
    booking_time: Optional[datetime] = None

    amount: float = Field(description="Amount as float for clean JSON output.")
    currency: str

    direction: TransactionDirection
    status: TransactionStatus

    country_code: Optional[str] = None
    merchant_category: Optional[str] = None
    channel: Optional[str] = None
    description: Optional[str] = None

    created_at: datetime
    updated_at: datetime

    # ---------- Normalizers for ORM values ----------

    @field_validator("amount", mode="before")
    def normalize_amount(cls, v: Any) -> float:
        """
        DB gives Decimal, we want a plain float in the response model.
        """
        if v is None:
            return 0.0
        return float(v)

    @field_validator("direction", mode="before")
    def normalize_direction(cls, v: Any) -> TransactionDirection:
        """
        Convert DB enum / raw string into our shared enum.
        """
        # Already our enum
        if isinstance(v, TransactionDirection):
            return v

        # Any Enum (e.g. SQLAlchemy enum) → take its .value
        if isinstance(v, Enum):
            v = v.value

        # Expect "DEBIT" or "CREDIT"
        return TransactionDirection(v)

    @field_validator("status", mode="before")
    def normalize_status(cls, v: Any) -> TransactionStatus:
        """
        Convert DB enum / raw string into our shared enum.
        """
        if isinstance(v, TransactionStatus):
            return v

        if isinstance(v, Enum):
            v = v.value

        # Expect "PENDING", "POSTED", "REVERSED", "CANCELLED"
        return TransactionStatus(v)

    class Config:
        from_attributes = True  # enables ORM → response model


class CreateTransactionRequest(BaseModel):
    """
    Data required to create a new Transaction.
    These fields map directly to columns, except ID/timestamps.
    """

    tx_id: str
    account_id: str
    customer_id: Optional[str] = None

    invoice_id: Optional[int] = None

    tx_time: datetime
    booking_time: Optional[datetime] = None

    amount: float = Field(
        description="Amount as float, will be stored as numeric/decimal in the DB."
    )
    currency: str = Field(max_length=3)

    direction: TransactionDirection
    status: TransactionStatus = TransactionStatus.POSTED

    country_code: Optional[str] = None
    merchant_category: Optional[str] = None
    channel: Optional[str] = None
    description: Optional[str] = None

    @field_validator("amount", mode="before")
    def convert_amount(cls, v: Any) -> float:
        # Normalize to float so JSON like "250.0" / "250" / 250 all work cleanly
        return float(v)

    class Config:
        json_schema_extra = {
            "example": {
                "tx_id": "TX-001",
                "account_id": "ACC-123",
                "customer_id": "CUST-999",
                "invoice_id": None,
                "tx_time": "2025-01-01T12:00:00Z",
                "booking_time": None,
                "amount": 249.50,
                "currency": "USD",
                "direction": "DEBIT",
                "status": "POSTED",
                "country_code": "US",
                "merchant_category": "5411",
                "channel": "ONLINE",
                "description": "Grocery purchase",
            }
        }

File: apps/analyst-dashboard/src/lib/api/transactions.ts
// src/lib/api/transactions.ts
import { fetchJson } from "@/lib/api/http"

export type TransactionDirection = "DEBIT" | "CREDIT" | (string & {})
export type TransactionStatus = "PENDING" | "POSTED" | "REVERSED" | "CANCELLED" | (string & {})

export type TransactionItem = {
  id: number
  tenant_id?: number | null
  transaction_reference?: string | null
  tx_id?: string | null
  account_id: string
  counterparty_account?: string | null
  customer_id?: string | null
  invoice_id?: number | null
  timestamp?: string | null
  tx_time?: string | null
  amount: number
  currency: string
  direction: TransactionDirection
  status: TransactionStatus
  country_code?: string | null
  merchant_category?: string | null
  channel?: string | null
  description?: string | null
  created_at: string
  updated_at: string
}

export type TransactionsListResponse = { items: TransactionItem[]; count?: number }

function normalizeItem(raw: any): TransactionItem {
  return {
    ...raw,
    // amount: API returns string decimal, normalize to number
    amount: typeof raw.amount === "string" ? parseFloat(raw.amount) : (raw.amount ?? 0),
    // normalize timestamp fields — API returns "timestamp", panel uses both
    tx_time: raw.tx_time ?? raw.timestamp ?? null,
    timestamp: raw.timestamp ?? raw.tx_time ?? null,
    // normalize id fields — API returns transaction_reference, panel uses tx_id
    tx_id: raw.tx_id ?? raw.transaction_reference ?? String(raw.id),
    transaction_reference: raw.transaction_reference ?? raw.tx_id ?? null,
  }
}

function normalizeListResponse(input: any): TransactionsListResponse {
  // { items, count } — /api/v1 envelope format
  if (input && Array.isArray(input.items)) {
    const count = typeof input.count === "number" ? input.count : input.items.length
    return { items: input.items.map(normalizeItem), count }
  }
  // { value, Count } — legacy /transactions/ format (current)
  if (input && Array.isArray(input.value)) {
    const count = typeof input.Count === "number" ? input.Count : input.value.length
    return { items: input.value.map(normalizeItem), count }
  }
  // raw array fallback
  if (Array.isArray(input)) {
    return { items: input.map(normalizeItem), count: input.length }
  }
  return { items: [], count: undefined }
}

export async function getTransactions(params: {
  account_id?: string
  customer_id?: string
  limit?: number
  offset?: number
}): Promise<TransactionsListResponse> {
  const sp = new URLSearchParams()
  const account = params.account_id?.trim()
  const customer = params.customer_id?.trim()
  if (account) sp.set("account_id", account)
  if (customer) sp.set("customer_id", customer)
  if (typeof params.limit === "number") sp.set("limit", String(params.limit))
  if (typeof params.offset === "number") sp.set("offset", String(params.offset))
  const qs = sp.toString()
  const data = await fetchJson<any>(`/transactions/${qs ? `?${qs}` : ""}`)
  return normalizeListResponse(data)
}

export async function getTransaction(id: number): Promise<TransactionItem> {
  const raw = await fetchJson<any>(`/transactions/${id}`)
  return normalizeItem(raw)
}
